src/Http/Routing/src/RouteEndpointBuilder.cs

```diff
@@ -79,9 +80,50 @@ public override Endpoint Build()
             requestDelegate,
             RoutePattern,
             Order,
-            new EndpointMetadataCollection(Metadata),
+            CreateMetadataCollection(Metadata),
             DisplayName);
 
         return routeEndpoint;
     }
+
+    private static EndpointMetadataCollection CreateMetadataCollection(IList<object> metadata)
+    {
+        if (metadata is { Count: > 0 })
+        {
+            var hasCorsMetadata = false;
+            IHttpMethodMetadata? httpMethodMetadata = null;
+
+            // Before create the final collection we
+            // need to update the IHttpMethodMetadata if
+            // a CORS metadata is present
+            for (var i = 0; i < metadata.Count; i++)
+            {
+                // Not using else if since a metadata could have both
+                // interfaces.
+
+                if (metadata[i] is IHttpMethodMetadata methodMetadata)
+                {
+                    // Storing only the last entry
+                    // since the last metadata is the most significant.
+                    httpMethodMetadata = methodMetadata;
+                }
+
+                if (!hasCorsMetadata && metadata[i] is ICorsMetadata)
+                {
+                    // IEnableCorsAttribute, IDisableCorsAttribute and ICorsPolicyMetadata
+                    // are ICorsMetadata
+                    hasCorsMetadata = true;
+                }
+            }
+
+            if (hasCorsMetadata && httpMethodMetadata is not null)
+            {
+                // Since we found a CORS metadata we will update it
+                // to make sure the acceptCorsPreflight is set to true.
+                httpMethodMetadata.AcceptCorsPreflight = true;
```

Comment:
Metadata is supposed to be immutable. Rather than modifying the HTTP method metadata, have you considered checking for ICorsMetadata inside the HTTP method policy? If it's present, then go ahead with accepting preflight.

Reply:
Unrelated to this PR but while we are on subject of immutable metadata. `IAuthorizeData` has public setters for its properties. Wondering if it should be deprecated and removed though it would be a breaking change.

Reply:
Metadata doesn't have to be immutable, and it's not worth breaking changes. It's a pattern to try to encourage.

Reply:
> Metadata is supposed to be immutable. Rather than modifying the HTTP method metadata, have you considered checking for ICorsMetadata inside the HTTP method policy? If it's present, then go ahead with accepting preflight.

@JamesNK that was my first idea but after talking with @halter73 the suggestion to mutate the metadata sounded reasonable

Reply:
@halter73 I can easily make the the changes in Http method policy without need a public api, do you have any concerns?

Reply:
The only behavior that I don't like about not mutating the metadata, change only the policy, is that the `AcceptsCorsPreFlight` will never be true, so, there is no reason for it to exist.

Reply:
@JamesNK Interesting idea. I also don't love mutating the metadata. We had a conversation about how to avoid we tried to this earlier in https://github.com/dotnet/aspnetcore/pull/43077#discussion_r937276637 by replacing the metadata, but that has other issues of course.

Having the CORS middleware handle this itself is another approach that I like. My problem with it is that:

> The only behavior that I don't like about not mutating the metadata, change only the policy, is that the AcceptsCorsPreFlight will never be true, so, there is no reason for it to exist.

If we go this route, we need to obsolete `AcceptCorsPreflight` and tell people to add the appropriate attribute the the endpoints metadata instead. I don't have much of a preference. As you say, we do have other mutable metadata, so we're not breaking entirely new ground here.

I've already approved this PR and I still think it's okay as is, but if people want to obsolete `AcceptCorsPreflight` instead, I'm okay with that too. That will need to be a new API proposal, but we could probably get that approved over email this afternoon if we think it's worth it.

Reply:
I don't have enough time before code complete to think and test this through to offer a strong opinion 😢

Reply:
Ok, if not objections I am planning to merge the PR as is. It is broken since 3.x, so I feel like it is a must to have change.